### PR TITLE
Added support for timezone in timestamp data type

### PR DESCRIPTION
**Pull request by grajoria**
Include timezone hour and minute in date time calculation.
Changes made to all the date time function to support the same.
Updated tests to check timezone corretness.

Signed-off-by: Girjesh Rajoria <grajoria@redhat.com>

### CODE REVIEW

File: include/s3select_functions.h [hunk 1]
```diff
@@ -814,13 +822,20 @@ struct _fn_to_timestamp : public base_function
...
-      result->set_value(&new_ptime);
     #endif
 
+    int tz_hour = tz_hr;
+    if ((char)sign == '-')
+    {
+      tz_hour *= -1;
+    }
+
+    tmstmp = std::make_tuple(new_ptime, boost::posix_time::time_duration(tz_hour, tz_mn, 0), (char)tm_zone == 'Z');
```
Comment (galsalomon66) [inline include/s3select_functions.h:835]: should validate that timezone-hour are between -12 --> +14

**Comment by galsalomon66** [comment]
extract is missing of the following time-parts
timezone_hour , timezone_minute

File: include/s3select_oper.h [hunk 2]
```diff
@@ -422,7 +422,7 @@ class value
     int64_t num;
     char* str;//TODO consider string_view
     double dbl;
-    boost::posix_time::ptime* timestamp;
+    std::tuple<boost::posix_time::ptime, boost::posix_time::time_duration, bool>* timestamp;
```
Comment (galsalomon66) [inline include/s3select_oper.h:425]: its better to use typedef , it will ease readability and also future modifications

**Comment by galsalomon66** [comment]
the base class for date-functions ease the timezone modifications

it needs to add tests for date_diff for timestamp with timezone

File: include/s3select_functions.h (hunk 1 above)
Comment (grajoria) [reply include/s3select_functions.h:842]: Done

File: include/s3select_oper.h (hunk 2 above)
Comment (grajoria) [reply include/s3select_oper.h]: Done

**Comment by grajoria** [comment]
@galsalomon66 Please review the changes. Thanks

File: include/s3select_functions.h [hunk 3]
```diff
@@ -694,15 +708,17 @@ struct _fn_to_timestamp : public base_function
...
-
+  int tz_hour;
 
   bool datetime_validation()
   {
-    if (yr >= 1400 && yr <= 9999 && mo >= 1 && mo <= 12 && dy >= 1 && hr < 24 && mn < 60 && sc < 60 && tz_hr < 24 && tz_mn < 60)
+    if (yr >= 1400 && yr <= 9999 && mo >= 1 && mo <= 12 && dy >= 1 && hr < 24 && mn < 60 && sc < 60 && tz_hour <= 14 && tz_hour >= -12  && tz_mn < 60)
     {
+      if ( (tz_hour == -12 || tz_hour == 14) && tz_mn > 0)
+        return false;
```
Comment (galsalomon66) [inline include/s3select_functions.h:720]: :+1:

File: include/s3select.h [hunk 4]
```diff
@@ -678,7 +678,7 @@ struct s3select : public bsc::grammar<s3select>
 
       date_part = (bsc::str_p("year") | bsc::str_p("month") | bsc::str_p("day") | bsc::str_p("hour")  | bsc::str_p("minute") | bsc::str_p("second")) [BOOST_BIND_ACTION(push_date_part)];
 
-      date_part_extract = ((date_part) |  bsc::str_p("week"));
+      date_part_extract = ((date_part) |  bsc::str_p("week") | bsc::str_p("timezone_hour") | bsc::str_p("timezone_minute"));
```
Comment (galsalomon66) [inline include/s3select.h:681]: :+1:

File: include/s3select_oper.h [hunk 5]
```diff
@@ -413,6 +413,8 @@ struct binop_modulo
   }
 };
 
+typedef std::tuple<boost::posix_time::ptime, boost::posix_time::time_duration, bool> timestamp_t;
```
Comment (galsalomon66) [inline include/s3select_oper.h:416]: :+1:

File: test/s3select_test.cpp [hunk 6]
```diff
@@ -391,40 +391,84 @@ TEST(TestS3selectFunctions, date_diff)
...
+    s3select_res = run_s3select(input_query);
+    EXPECT_EQ(s3select_res, "-1052733");
+
     input_query = "select date_diff(second, to_timestamp(\'2009-09-17T17:56:06.234Z\'), to_timestamp(\'2009-09-17T19:30:05.234Z\')) from stdin;" ;
     s3select_res = run_s3select(input_query);
     EXPECT_EQ(s3select_res, "5639");
+
+    input_query = "select date_diff(hour, to_timestamp(\'2007-09-17T17:56:06.234-03:45\'), to_timestamp(\'2007-09-17T17:56:06.234+13:30\')) from stdin;" ;
+    s3select_res = run_s3select(input_query);
+    EXPECT_EQ(s3select_res, "-17");
```
Comment (galsalomon66) [inline test/s3select_test.cpp:412]: result should be 17:15?

File: test/s3select_test.cpp [hunk 7]
```diff
@@ -391,40 +391,84 @@ TEST(TestS3selectFunctions, date_diff)
...
 
+    input_query = "select date_diff(minute, to_timestamp(\'2009-09-17T19:30:05.234Z\'), to_timestamp(\'2007-09-17T17:56:06.234Z\')) from stdin;" ;
+    s3select_res = run_s3select(input_query);
+    EXPECT_EQ(s3select_res, "-1052733");
+
     input_query = "select date_diff(second, to_timestamp(\'2009-09-17T17:56:06.234Z\'), to_timestamp(\'2009-09-17T19:30:05.234Z\')) from stdin;" ;
     s3select_res = run_s3select(input_query);
     EXPECT_EQ(s3select_res, "5639");
+
+    input_query = "select date_diff(hour, to_timestamp(\'2007-09-17T17:56:06.234-03:45\'), to_timestamp(\'2007-09-17T17:56:06.234+13:30\')) from stdin;" ;
```
Comment (galsalomon66) [inline test/s3select_test.cpp:410]: :+1:

File: test/s3select_test.cpp [hunk 8]
```diff
@@ -391,40 +391,84 @@ TEST(TestS3selectFunctions, date_diff)
...
+    s3select_res = run_s3select(input_query);
+    EXPECT_EQ(s3select_res, "14");
+
+    input_query = "select date_diff(minute, to_timestamp(\'2007-09-17T17:56:06.234-03:45\'), to_timestamp(\'2007-09-17T17:56:06.234+13:30\')) from stdin;" ;
+    s3select_res = run_s3select(input_query);
+    EXPECT_EQ(s3select_res, "-1035");
+
+    input_query = "select date_diff(minute, to_timestamp(\'2007-09-17T17:56:06.234+03:45\'), to_timestamp(\'2007-09-17T17:56:06.234+13:30\')) from stdin;" ;
+    s3select_res = run_s3select(input_query);
+    EXPECT_EQ(s3select_res, "-585");
```
Comment (galsalomon66) [inline test/s3select_test.cpp:432]: :+1:

**Review by galsalomon66 — APPROVED** [review]

**Pull request merged**